cpp/include/cudf/join.hpp

```diff
@@ -367,5 +367,138 @@ std::unique_ptr<cudf::table> cross_join(
   cudf::table_view const& right,
   rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource());
 
+/**
+ * @brief Hash join that builds hash table in creation and probes results in subsequent `*_join`
+ * member functions.
+ *
+ * This class enables the hash join scheme that builds hash table once, and probes as many times as
+ * needed (possibly in parallel).
+ */
+class hash_join {
+ public:
+  hash_join() = delete;
+  ~hash_join();
+  hash_join(hash_join const&) = delete;
+  hash_join(hash_join&&)      = delete;
+  hash_join& operator=(hash_join const&) = delete;
+  hash_join& operator=(hash_join&&) = delete;
+
+  /**
+   * @brief Construct a hash join object and build the internal hash table used for subsequent
+   * probe calls.
+   *
+   * @param build The build table, from which the hash table is built.
+   * @param build_on The column indices from `build` to join on.
+   */
+  hash_join(cudf::table_view const& build, std::vector<size_type> const& build_on);
+
+  /**
+   * @brief Side of the probe table in the joined table. Only applicable for inner join.
+   */
+  enum class probe_output_side { LEFT, RIGHT };
+
+  /**
+   * @brief Performs an inner join by probing in the internal hash table.
+   *
+   * Given that it is sometimes desired to choose the small table to be the `build` side for an
+   * inner join (which is automatically done in cudf::innner_join()), it is handy to be able to
+   * specify which side in the joined table the `probe` portion is on. That is, one can check
+   * whether to use `left` or `right` table to be the build side by comparing their row counts ahead
+   * of `hash_join` object, but still preserve the logical order of the `left`/`right` in the joined
+   * table. For instance, if the `left` table is the small one (and thus the `build` side),
+   * specifying the `probe_output_side` to `RIGHT` will output the `right` table (probe side) on the
+   * right side in the joined table.
+   *
+   * More details please @see cudf::inner_join().
+   *
+   * @param probe The probe table, from which the tuples are probed.
+   * @param probe_on The column indices from `probe` to join on.
+   * @param columns_in_common is a vector of pairs of column indices into
+   * `build` and `probe`, respectively, that are "in common". For "common"
+   * columns, only a single output column will be produced, which is gathered
+   * from `probe_on` columns or `build_on` columns if `probe_output_side` is LEFT or RIGHT.
+   * Else, for every column in `probe_on` and `build_on`,
+   * an output column will be produced. For each of these pairs (P, B), P
+   * should exist in `probe_on` and B should exist in `build_on`.
```

Comment:
I really dislike this whole `columns_in_common` business we had in the old API. The whole point of it is to avoid unnecessarily materializing columns that Pandas wouldn't materialize because the columns have the same name. Don't worry about it in this PR, but I think we need to find a better way to do this. Perhaps by having APIs that just return the indices of matching rows and then Python can handle gathering from the appropriate columns.